packages/muelu/src/Graph/UncoupledAggregation/MueLu_AggregationPhase2bAlgorithm_kokkos_def.hpp

```diff
@@ -90,57 +114,85 @@ namespace MueLu {
-        }
-
-        int bestScore   = -100000;
-        int bestAggId   = -1;
-        int bestConnect = -1;
-
-        for (int j = 0; j < as<int>(neighOfINode.length); j++) {
-          LO neigh = neighOfINode(j);
-
-          if (graph.isLocalNeighborVertex(neigh) && aggStat[neigh] == AGGREGATED) {
-            int aggId = vertex2AggId[neigh];
-            int score = aggWeight[aggId] - aggPenalties[aggId];
-
-            if (score > bestScore) {
-              bestAggId   = aggId;
-              bestScore   = score;
-              bestConnect = connectWeight[neigh];
-
-            } else if (aggId == bestAggId && connectWeight[neigh] > bestConnect) {
-              bestConnect = connectWeight[neigh];
-            }
-
-            // Reset the weights for the next loop
-            aggWeight[aggId] = 0;
-          }
-        }
-
-        if (bestScore >= 0) {
-          aggStat     [i] = AGGREGATED;
-          vertex2AggId[i] = bestAggId;
-          procWinner  [i] = myRank;
-
-          numNonAggregatedNodes--;
+      // total work = numberOfTeams * teamSize
+      Kokkos::TeamPolicy<execution_space> outerPolicy(numRows, Kokkos::AUTO);
+      typedef typename Kokkos::TeamPolicy<execution_space>::member_type  member_type;
+      // Kokkos::RangePolicy<execution_space> numRowsPolicy(0, numRows);
+      LO tmpNumNonAggregatedNodes = 0;
+      Kokkos::parallel_reduce("Aggregation Phase 2b: aggregates expansion",
+                              // numRowsPolicy,
+                              outerPolicy.set_scratch_size( 0, Kokkos::PerTeam( scratch_size ) ),
+                              KOKKOS_LAMBDA (const member_type &teamMember,
+                                             LO& lNumNonAggregatedNodes) {
+
+                                // Retrieve the id of the vertex we are currently working on and
+                                // allocate view locally so that threads do not trash the weigth
+                                // when working on the same aggregate.
+                                const int vertexIdx = teamMember.league_rank();
+                                ScratchViewType aggWeightView(teamMember.team_scratch( 0 ),
+                                                              maxNumNeighbors);
+
+                                if (aggStatView(vertexIdx) == READY) {
+
+                                  auto neighOfINode = graph.getNeighborVertices(vertexIdx);
+
+                                  for (int j = 0; j < as<int>(neighOfINode.length); j++) {
+                                    LO neigh = neighOfINode(j);
+
+                                    // We don't check (neigh != i), as it is covered by checking
+                                    // (aggStat[neigh] == AGGREGATED)
+                                    if ( graph.isLocalNeighborVertex(neigh)
+                                         && (aggStatView(neigh) == AGGREGATED) )
+                                      aggWeightView(j) = aggWeightView(j)
```

Comment:
Index j is wrong.  Instead, we want to the weight to be computed per aggregate, not per individual nbr.